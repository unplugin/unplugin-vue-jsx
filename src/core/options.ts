import process from 'node:process'
import { getPackageInfo } from 'local-pkg'
import type { VueJSXPluginOptions } from '@vue/babel-plugin-jsx'
import type { FilterPattern } from '@rollup/pluginutils'

export type Vue2JSXOptions = {
  functional?: boolean
  injectH?: boolean
  vModel?: boolean
  vOn?: boolean
  compositionAPI?: 'auto' | 'native' | 'plugin' | boolean | 'naruto'
}

export type Options = {
  include?: FilterPattern
  exclude?: FilterPattern | undefined
  root?: string
  sourceMap?: boolean
} & (OptionsVue2 | OptionsVue3 | { version?: 'auto' })

interface OptionsVue2 extends Vue2JSXOptions {
  version: 2
}

interface OptionsVue3 extends VueJSXPluginOptions {
  version: 3
}

export type OptionsResolved = Omit<Required<Options>, 'version'> & {
  version: 2 | 3
}

export async function resolveOption(
  options: Options,
): Promise<OptionsResolved> {
  const root = options.root || process.cwd()
  let version: 2 | 3
  if (options.version === 'auto') {
    version = await getVueVersion(root)
  } else version = options.version || 3

  return {
    ...options,
    include: options.include || [/\.[jt]sx?$/],
    exclude: options.exclude || undefined,
    version,
    root,
    sourceMap: options.sourceMap ?? true,
  }
}

async function getVueVersion(root: string) {
  const pkg = await getPackageInfo('vue', { paths: [root] })
  if (!pkg) return 3
  return pkg.version?.startsWith('2') ? 2 : 3
}
