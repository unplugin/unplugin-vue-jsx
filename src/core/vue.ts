import { transform, type TransformOptions } from '@babel/core'
import BabelVueJsx from '@vue/babel-plugin-jsx'
import type { OptionsResolved } from './options'

function isTS(id: string): boolean {
  return /\.[cm]?tsx?$/.test(id)
}

export function transformVueJsx(
  code: string,
  id: string,
  options: Omit<OptionsResolved, 'include' | 'exclude'>,
): { code: string; map: any } | undefined {
  const tsSyntax = isTS(id)

  const transformOptions: TransformOptions = {
    babelrc: false,
    configFile: false,
    plugins: [[BabelVueJsx, options], ...options.babelPlugins],
    sourceMaps: options.sourceMap,
    sourceFileName: id,
    parserOpts: {
      ...options.parserOpts,
      plugins: [
        ...(tsSyntax ? (['typescript', 'jsx'] as const) : []),
        ...(options.parserOpts?.plugins || []),
      ],
    },
  }

  const result = transform(code, transformOptions)
  if (!result?.code) return

  return {
    code: result.code,
    map: result.map,
  }
}
