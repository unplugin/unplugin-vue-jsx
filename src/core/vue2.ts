import { type TransformOptions, transform } from '@babel/core'
// @ts-expect-error
import TS from '@babel/plugin-syntax-typescript'
// @ts-expect-error
import vue2Jsx from '@vue/babel-preset-jsx'
import { isTS } from './utils'
import type { OptionsResolved } from './options'

export const transformVue2 = (
  code: string,
  id: string,
  options: OptionsResolved,
) => {
  const transformOptions: TransformOptions = {
    babelrc: false,
    configFile: false,
    plugins: [],
    presets: [[vue2Jsx, options]],
    sourceMaps: options.sourceMap,
    sourceFileName: id,
  }

  if (isTS(id)) {
    transformOptions.plugins!.push([TS, { isTSX: true }])
  }

  const result = transform(code, transformOptions)
  if (!result?.code) return

  return {
    code: result.code,
    map: result.map,
  }
}
