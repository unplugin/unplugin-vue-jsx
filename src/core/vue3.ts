import { type TransformOptions, transform } from '@babel/core'
import vue3Jsx from '@vue/babel-plugin-jsx'
// @ts-expect-error
import TS from '@babel/plugin-syntax-typescript'
import { isTS } from './utils'
import type { OptionsResolved } from './options'

export const transformVue3 = (
  code: string,
  id: string,
  options: OptionsResolved,
) => {
  const transformOptions: TransformOptions = {
    babelrc: false,
    configFile: false,
    plugins: [[vue3Jsx, options]],
    sourceMaps: options.sourceMap,
    sourceFileName: id,
  }

  if (isTS(id)) {
    transformOptions.plugins!.push([TS, { isTSX: true }])
  }

  const result = transform(code, transformOptions)
  if (!result?.code) return

  return {
    code: result.code,
    map: result.map,
  }
}
