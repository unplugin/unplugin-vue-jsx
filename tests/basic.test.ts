import { describe, expect, test } from 'vitest'
import { resolveOptions, type Options } from '../src/core/options'
import { transformVueJsx } from '../src/core/vue'

function transform(code: string, isTS = false, userOptions: Options = {}) {
  const options = resolveOptions(userOptions)
  return transformVueJsx(code, `foo.${isTS ? 'tsx' : 'jsx'}`, options)?.code
}

describe('Vue 3', () => {
  test('basic', () => {
    expect(transform('<div />')).toMatchSnapshot()
    expect(transform('<div key="1" />')).toMatchSnapshot()
    expect(transform("<div foo={'bar'} />")).toMatchSnapshot()
  })

  test('typescript', () => {
    expect(transform('const foo: any = <div />', true)).toMatchSnapshot()
  })

  test('custom options', () => {
    expect(
      transform(`<input on={{ click: a }} />`, false, {
        transformOn: false,
      }),
    ).toMatchSnapshot()
    expect(
      transform(`<input on={{ click: a }} />`, false, {
        transformOn: true,
      }),
    ).toMatchSnapshot()
  })
})
