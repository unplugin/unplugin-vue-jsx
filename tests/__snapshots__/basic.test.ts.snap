// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Vue 3 > basic 1`] = `
"import { createVNode as _createVNode } from "vue";
_createVNode("div", null, null);"
`;

exports[`Vue 3 > basic 2`] = `
"import { createVNode as _createVNode } from "vue";
_createVNode("div", {
  "key": "1"
}, null);"
`;

exports[`Vue 3 > basic 3`] = `
"import { createVNode as _createVNode } from "vue";
_createVNode("div", {
  "foo": 'bar'
}, null);"
`;

exports[`Vue 3 > custom options 1`] = `
"import { createVNode as _createVNode } from "vue";
_createVNode("input", {
  "on": {
    click: a
  }
}, null);"
`;

exports[`Vue 3 > custom options 2`] = `
"import { createVNode as _createVNode } from "vue";
import _transformOn from "@vue/babel-helper-vue-transform-on";
_createVNode("input", _transformOn({
  click: a
}), null);"
`;

exports[`Vue 3 > typescript 1`] = `
"import { createVNode as _createVNode } from "vue";
const foo: any = _createVNode("div", null, null);"
`;
